File: pumpkin-lib/examples/nqueens.rs
use pumpkin_lib::constraints;
use pumpkin_lib::results::ProblemSolution;
use pumpkin_lib::results::SatisfactionResult;
use pumpkin_lib::termination::Indefinite;
use pumpkin_lib::variables::TransformableVariable;
use pumpkin_lib::Solver;

fn main() {
    let n = std::env::args()
        .nth(1)
        .expect("Please provide a value for 'n'")
        .parse::<u32>()
        .expect("'n' is not a valid unsigned integer");

    if n < 2 {
        println!("Please provide an 'n > 1'");
        return;
    }

    let mut solver = Solver::default();
    let variables = (0..n)
        .map(|_| solver.new_bounded_integer(0, n as i32 - 1))
        .collect::<Vec<_>>();

    let _ = solver
        .add_constraint(constraints::all_different(variables.clone()))
        .post();

    let diag1 = variables
        .iter()
        .cloned()
        .enumerate()
        .map(|(i, var)| var.offset(i as i32))
        .collect::<Vec<_>>();
    let diag2 = variables
        .iter()
        .cloned()
        .enumerate()
        .map(|(i, var)| var.offset(-(i as i32)))
        .collect::<Vec<_>>();

    let _ = solver
        .add_constraint(constraints::all_different(diag1))
        .post();
    let _ = solver
        .add_constraint(constraints::all_different(diag2))
        .post();

    let mut brancher = solver.default_brancher_over_all_propositional_variables();
    match solver.satisfy(&mut brancher, &mut Indefinite) {
        SatisfactionResult::Satisfiable(solution) => {
            let row_separator = format!("{}+", "+---".repeat(n as usize));

            for row in 0..n {
                println!("{row_separator}");

                let queen_col = solution.get_integer_value(variables[row as usize]) as u32;

                for col in 0..n {
                    let string = if queen_col == col { "| * " } else { "|   " };

                    print!("{string}");
                }

                println!("|");
            }

            println!("{row_separator}");
        }
        SatisfactionResult::Unsatisfiable => {
            println!("{n}-queens is unsatisfiable.");
        }
        SatisfactionResult::Unknown => {
            println!("Timeout.");
        }
    }
}

File: pumpkin-py/python/examples/nqueens.py
from argparse import ArgumentParser
from pathlib import Path
from typing import Optional

from pumpkin_py import constraints, SatisfactionResult, Solver

def main(n: int, proof: Path | None):
    assert n > 0, "Please provide a positive non-zero 'n'"

    solver = Solver(proof=proof)

    variables = [solver.new_variable(0, n - 1, name=f"q{i}") for i in range(n)]

    solver.post(constraints.all_different(variables))

    diag1 = [var.offset(i) for (i, var) in enumerate(variables)]
    diag2 = [var.offset(-i) for (i, var) in enumerate(variables)]

    solver.post(constraints.all_different(diag1))
    solver.post(constraints.all_different(diag2))

    status = solver.satisfy()
    match status:
        case SatisfactionResult.Satisfiable(solution):
            row_separator = "+---" * n + "+"

            for row in range(n):
                print(f"{row_separator}");

                queen_col = solution.value(variables[row])

                for col in range(n): 
                    string = "| * " if queen_col == col else "|   "
                    print(f"{string}", end='')
                
                print("|")
            
            print(f"{row_separator}")
        
        case SatisfactionResult.Unsatisfiable():
            print(f"{n}-queens is unsatisfiable.")
        
        case SatisfactionResult.Unknown():
            print("Timeout.")


if __name__ == "__main__":
    arg_parser = ArgumentParser()

    arg_parser.add_argument("--proof", type=Path, help="The proof file.", default=None)
    arg_parser.add_argument("n", type=int, help="The size of the chessboard.")

    args = arg_parser.parse_args()

    main(args.n, args.proof)
